File: calc_aa.py
import numpy as np
from Get_global_value import num_q
from Get_global_value import BB
from Get_global_value import J_type
from Get_global_value import Qi
from rpy2dc import rpy2dc


def calc_aa(A0, q):
    AA = np.zeros((num_q, 3, 3))
    if num_q == 0:
        print('Single body, there is no link')

    else:
        A_I_0 = A0

        for i in range(num_q):
            if BB[i] == -1:
                if J_type[i] == 'R':
                    A_0_i = rpy2dc(Qi[i, 0], Qi[i, 1], Qi[i, 2] + q[i]).T
                else:
                    A_0_i = rpy2dc(Qi[i, :], 0, 0).T

                AA[i, 0:3, 0:3] = A_I_0 @ A_0_i

            else:
                if J_type[i] == 'R':
                    A_BB_i = rpy2dc(Qi[i, 0], Qi[i, 1], Qi[i, 2] + q[i]).T
                else:
                    A_BB_i = rpy2dc(Qi[i, 0:3], 0, 0).T

                AA[i, 0:3, 0:3] = AA[BB[i], 0:3, 0:3] @ A_BB_i

    return AA

File: src/matlab/spacedyn_v2r1/calc_aa.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% The SpaceDyn, a MATLAB toolbox for Space and Mobile Robots.
% (C)1998 The Space Robotics Lab. directed by Kazuya Yoshida,
% Tohoku University, Japan.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%   version 1 // Oct 4, 1999, Last modification by K.Yoshida
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
% CALC_AA 	Calculate the coordinate transfrom matrices in the Robotics convention.
%
%   CALC_TRN( A0 , q ) returns the coordinate tranformation matrices AA.
%   AA is a collection of A_I_1, A_I_2, ... A_I_n.
%
%   2002.2.27 H.Hamano modified to version up
%

function SV = calc_aa( LP, SV )


% If a Single body,
if LP.num_q == 0
   
   SV.AA = [];
   
% If a Multi body system,
else
   
   % Calculation of coordinate transformation matrices
   A_I_0 = SV.A0;
   
   for i = 1 : LP.num_q
      
      % Check the link connection: Is the lower one of this link, 0 ?
      if LP.BB(i) == 0
         
         % Current (i-th) link connects to the 0-th link
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            A_0_i = (rpy2dc( LP.Qi(1,i) , LP.Qi(2,i) , LP.Qi(3,i)+SV.q(i) )');
            
         % Prismatic joint
         else
            A_0_i = (rpy2dc( LP.Qi(:,i) )');
            
         end
         
         SV.AA(:,i*3-2:i*3) = A_I_0*A_0_i;
         
      else
         
         % Current (i-th) link doesn't connect to the 0-th link
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            A_BB_i = (rpy2dc( LP.Qi(1,i), LP.Qi(2,i), LP.Qi(3,i)+SV.q(i) )');
            
         % Prismatic joint
         else
            A_BB_i = (rpy2dc( LP.Qi(:,i) )');
            
         end
         
         SV.AA(:,i*3-2:i*3) = SV.AA(:,LP.BB(i)*3-2:LP.BB(i)*3)*A_BB_i;
         
      end
      
   end
   
end


%%%EOF
